lib/AST/ASTStructuralEquivalence.cpp

```diff
@@ -995,6 +995,17 @@ static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
         D1CXX->getASTContext().getExternalSource()->CompleteType(D1CXX);
       }
 
+      if (auto *T1 = D1CXX->getDescribedClassTemplate()) {
```

Comment:
Perhaps instead of having so many nested if-else we could have a function which tests template incompatibility. Something like the below draft:
```
bool areCompatibleTemplates(CXXRecordDecl *D1CXX, CXXRecordDecl *D2CXX) {
  auto Desc1 = D1CXX->getDescribedClassTemplate();
  auto Desc2 = D2CXX->getDescribedClassTemplate();
 return (Desc1 && Desc2) || (!Desc1 && !Desc2); 
}
```
Then we could use it like this:
```
if (!areCompatibleTemplates(D1CXX, D2CXX))
  return false;
if (!IsStructurallyEquivalent(Context, ...))
  return false;
```

What do you think?